### PR TITLE
TREX-14 Eligibility & Compliance checker contract

### PR DESCRIPTION
**Pull request by pgonday**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -85,5 +85,8 @@
     "*.js": [
       "eslint"
     ]
+  },
+  "dependencies": {
+    "hardhat-tracer": "^3.0.3"
```
Comment (Nakasar) [inline package.json:90]: Should be in `devDependencies`

```bash
npm uninstall hardhat-tracer && npm install --save-dev hardhat-tracer
```

File: contracts/utilities/IUtilityChecker.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    ///     - check if token is paused,
+    ///     - check freeze conditions,
+    ///     - check compliance.
+    /// @param _compliance Address of the Modular Compliance contract.
+    /// @param _token The address of the token contract.
+    /// @param _from The address of the sender.
+    /// @param _to The address of the recipient.
+    /// @param _amount The amount of tokens to be transferred.
+    /// @return bool Returns true if the transfer would be successful, false otherwise.
+    function testTransfer(address _compliance, address _token, address _from, address _to, uint256 _amount)
```
Comment (Joachim-Lebrun) [inline contracts/utilities/IUtilityChecker.sol:101]: the interface can be simplified by removing the compliance address, this address can be found easily if you have the token address as there is a getter on the token interface to get the compliance address.

**Comment by Joachim-Lebrun** [comment]
@pgonday can you solve the conflicts on the PR please ?

File: contracts/utilities/UtilityChecker.sol [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+contract UtilityChecker is IUtilityChecker, OwnableUpgradeable, UUPSUpgradeable {
+
+    function initialize() external initializer {
+        __Ownable_init();
+    }
+
+    /// @inheritdoc IUtilityChecker
+    /// @dev This function is not gas optimized and should be called only OFF chain.
+    function testTransfer(address _token, address _from, address _to, uint256 _amount) 
+        external view override returns (bool) {
```
Comment (Joachim-Lebrun) [inline contracts/utilities/UtilityChecker.sol:86]: we could make this function return 3 boolean values 
```suggestion
        external view override returns (bool _freezeStatus, bool _eligibilityStatus, bool _complianceStatus) {
```
This way the user knows immediately where to check to get the explanation of why the transfer is going to fail

File: contracts/utilities/UtilityChecker.sol [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        IToken token = IToken(_token);
+
+        if (token.paused()) return false;
+
+        (bool frozen, ) = testFreeze(_token, _from, _to, _amount);
+        if (frozen) return false;
+
+        ComplianceCheckDetails [] memory details = testTransferDetails(_token, _from, _to, _amount);
```
Comment (Joachim-Lebrun) [inline contracts/utilities/UtilityChecker.sol:94]: you check compliance, pause and freeze but not the eligibility in this function, which means that right now a transfer that would block at the eligibility level will return true, while it should return false. You need to check the eligibility on the receiver address (just need to call `isVerified` on the IR of the token

Comment (Joachim-Lebrun) [reply contracts/utilities/UtilityChecker.sol:94]:
```suggestion
        IIdentityRegistry ir = token.identityRegistry();
        if(!ir.isVerified(_to)) return false; 
        ComplianceCheckDetails [] memory details = testTransferDetails(_token, _from, _to, _amount);
```

**Review by Joachim-Lebrun — CHANGES_REQUESTED** [review]

**Comment by pgonday** [comment]
Relaced by https://github.com/TokenySolutions/T-REX/pull/229

**Pull request closed without merging**